coil-video/src/main/java/coil/fetch/MediaDataSourceFetcher.kt

```diff
@@ -0,0 +1,84 @@
+package coil.fetch
+
+import android.media.MediaDataSource
+import android.os.Build
+import androidx.annotation.RequiresApi
+import coil.ImageLoader
+import coil.annotation.ExperimentalCoilApi
+import coil.decode.DataSource
+import coil.decode.ImageSource
+import coil.request.Options
+import okio.Buffer
+import okio.Source
+import okio.Timeout
+import okio.buffer
+
+@RequiresApi(Build.VERSION_CODES.M)
+@OptIn(ExperimentalCoilApi::class)
+class MediaDataSourceFetcher(
+    private val data: MediaDataSource,
+    private val options: Options,
+) : Fetcher {
+
+    override suspend fun fetch(): FetchResult {
+        val imageSource = ImageSource(
+            source = MediaDataSourceOkIoSource(data).buffer(),
+            context = options.context,
+            metadata = MediaSourceMetadata(data),
+        )
+
+        return SourceResult(
+            source = imageSource,
+            mimeType = "video/", // TODO: How to get the mime type?
```

Comment:
I had this `null` before but then `VideoFrameDecoder.isApplicable()` would return `false`. When setting it `null` how to still pass it down to the video decoder?

> Technically the MediaDataSource could even be reading from a non-video source file.

I think at least audio files are also intended for `MediaDataSource`, however the interface is that generic every file could be wrapped in it.

Reply:
It's possible to force a custom decoder to be used for a specific `ImageRequest` using [this method](https://coil-kt.github.io/coil/api/coil-base/coil.request/-image-request/-builder/decoder-factory.html?query=fun%20decoderFactory(factory:%20Decoder.Factory):%20ImageRequest.Builder). That's how `VideoFrameDecoder` is intended to be used today.

Long term the solution is to update `VideoFrameDecoder` to read a stream's data to determine if it's a supported type instead of relying on the MIME type. Though, this is a lot of work.

Reply:
Thanks, wasn't aware of that method actually. Changed the `mimeType` to `null`.